File: client/lib/data/models/historicalposition.dart
import 'package:enum_to_string/enum_to_string.dart';
import 'package:solidtrade/data/common/shared/buy_or_sell.dart';
import 'package:solidtrade/data/common/shared/position_type.dart';
import 'package:solidtrade/data/models/base_entity.dart';

class HistoricalPosition implements IBaseEntity {
  @override
  final int id;
  @override
  final DateTime createdAt;
  @override
  final DateTime updatedAt;

  final String isin;
  final double buyInPrice;
  final double performance;
  final double numberOfShares;
  final BuyOrSell buyOrSell;
  final PositionType positionType;

  const HistoricalPosition({
    required this.id,
    required this.createdAt,
    required this.updatedAt,
    required this.positionType,
    required this.buyOrSell,
    required this.buyInPrice,
    required this.isin,
    required this.numberOfShares,
    required this.performance,
  });

  factory HistoricalPosition.fromJson(Map<String, dynamic> json) {
    return HistoricalPosition(
      id: json["id"],
      createdAt: DateTime.parse(json["createdAt"]),
      updatedAt: DateTime.parse(json["updatedAt"]),
      positionType: EnumToString.fromString(PositionType.values, json["positionType"])!,
      buyOrSell: EnumToString.fromString(BuyOrSell.values, json["buyOrSell"])!,
      buyInPrice: json["buyInPrice"],
      performance: json["performance"],
      numberOfShares: json["numberOfShares"],
      isin: json["isin"],
    );
  }

  factory HistoricalPosition.copy(HistoricalPosition hp, double numberOfShares) {
    return HistoricalPosition(
      id: hp.id,
      buyInPrice: hp.buyInPrice,
      buyOrSell: hp.buyOrSell,
      createdAt: hp.createdAt,
      isin: hp.isin,
      numberOfShares: numberOfShares,
      performance: hp.performance,
      positionType: hp.positionType,
      updatedAt: hp.updatedAt,
    );
  }
}

File: server/Domain/Entities/HistoricalPosition.cs
using System.ComponentModel.DataAnnotations;
using Domain.Common;
using Domain.Common.Position;
using Domain.Entities.Base;
using Domain.Enums;

namespace Domain.Entities;

public class HistoricalPosition : BaseEntity, IPosition
{
    public int UserId { get; set; }
        
    [Required]
    public PositionType PositionType { get; set; }
        
    [Required]
    public BuyOrSell BuyOrSell { get; set; }
        
    [Required]
    public decimal BuyInPrice { get; set; }
        
    [Required]
    public decimal Performance { get; set; }
        
    [Required]
    [Range(DomainConstants.MinimumNumberOfShares, int.MaxValue)]
    public decimal NumberOfShares { get; set; }
        
    [Required]
    [MaxLength(12)]
    public string Isin { get; set; }
}